### ISSUE TITLE
Android - Intercom widget should never be shawn before deviceReady has ended

### ISSUE THREAD

**Issue by slorber** [intercom/intercom-cordova#18]
It may apply to iOS too but I don't have any iPhone to test

--------------------------------------

# Summary

When the app is starting in a logged out state: it may, at the very beginning, show the intercom widget that corresponds to the last connected user (instead of showing nothing)

When a logged in user kills the application (or the application crash) and restarts it in logged-out state, there's a small amount of time where the intercom widget may be shawn.
In my case it is even shawn on top of the Splash Screen, before the onDeviceReady event gets fired (because my JS takes some time to parse)

---------------------------------

# Full

I have an app with a login screen, and only logged-in user.
I mean I don't use `intercom.registerUnidentifiedUser()` because the logged-out part is very small.

The user may or not want to keep being logged in on app restart (ie a checkbox "remember me" option during login).

When the user log in, I call `intercom.registerIdentifiedUser(intercomData);`

My app startup pseudo-code looks like this:

```javascript
function deviceReady() {
    hideSplashScreen();
    if ( autoConnectUser ) {
        intercom.registerIdentifiedUser(userIntercomData);
        startApp();
    } 
    else {
        intercom.reset();
        showLoginForm();
    }
}
```
My JS application is heavy. The parsing of the JS files on a bad device can take something like 3 seconds.

When the app is killed, it seems the intercom registered user is not removed (probably on purpose).

When the app is restarted in logged out state (because the user did not check the "remember me" checkbox), then there's a small amount of time where the intercom widget shows up, and then it is removed (because my `deviceReady` call `reset()`). This somehow corresponds to the time my JS takes to be parsed, and the widget appears on top of the splash screen.

I think the intercom plugin should never display the widget before the deviceReady function has ended, giving the oppurtunity to the developer to handle "implicit logouts" that can be caused by application being killed, application crash... (we can't listen for kill/crash events)

Also I'm not sure if it's worth to allow the intercom widget to be shawn on top of the splash screen.

**Comment by harkin** [comment]
hi @slorber 

You can tell the library when to show and when not to show the widget using the `setVisibility` method. Calling it with `GONE` will stop it appearing and then when you're ready for it to appear call it with `VISIBLE`. This method caches the last value it was given and the default value if you've never set it is visible. This does mean that if it's `VISIBLE` and the app crashes and restarts it will still be `VISIBLE`, likewise with `GONE`. 

I don't know the lifecycle of cordova apps _at all_. Is it possible for you to call that method in some setup method for your Splashscreen or when the app is opened with the `GONE ` parameter? Then change it to `VISIBLE` when the user logs in or you do the logic about logging them in if they've ticked 'remember me'.

**Comment by harkin** [comment]
As an aside, I'm not sure you'll have correct behaviour with how you use `reset()`. You'll almost certainly see session counts being inflated beyond what they should be. If you're in a state where you remember who the user is you shouldn't call `reset()` and for the state where when the app closes you forget who that user is you should try to call `reset()` on app close and not on next app launch.

**Comment by slorber** [comment]
@harkin we are not supposed to access cordova/plugins API before the `ondeviceready` event.
So the first opportunity I have to call `intercom.setVisibility(intercom.GONE);` is in the `deviceReady` function. But unfortunatly the intercom widget is already displayed at this point :( Even if the first thing I do is hide intercom, it still gets displayed during app initialization.

Also note that I understand the idea of calling `reset()` on app close/kill/crash. But there is [no way](http://stackoverflow.com/questions/25109820/whats-the-event-when-an-app-exits-in-cordova) at all to detect when app is being closed and run code at this moment.


Notice I tried the following and it did not work: the intercom widget is still coming on top of the splash screen, before the login form is displayed.

```javascript
function deviceReady() {

     // Never show the intercom widget for the 15 first seconds on app startup 
    intercom.setVisibility(intercom.GONE);
    setTimeout(function() {
        intercom.setVisibility(intercom.VISIBLE);
    },15000);

    hideSplashScreen();
    if ( autoConnectUser ) {
        intercom.registerIdentifiedUser(userIntercomData);
        startApp();
    } 
    else {
        intercom.reset();
        showLoginForm();
    }
}
```



--------------------------------------

About using `reset()` everytime on app restarts, actually it does not really matter to me if it increments the number of sessions. Otherwise I would often end up with users that may be very active but intercom only displays 1 session, because the users never logout. Intercom does not seem to compute the "usage time" of the app, so I prefer incrementing the session counter everytime the app restarts (or maybe i'll use an event in the future...). Also notice I think it mimics the session counter usage of our desktop single page app.

**Comment by harkin** [comment]
Ah that's certainly not ideal, @jtreanor do you have any ideas for stopping the widget showing up?

Not knowing when the app is closed sucks, we've faced the same issue on Android and come up with a reasonably good solution that involves keeping count of how many activities we know are in an active state. If the number becomes 0 we assume the app is going into the background. 

To clarify what I meant about session count being inflated we increment the session count by 1 under these circumstances:
1. when you register a user
2. when you open an app with a user already registered if:
    * the app has been in the background for more than 20 seconds
    * the app was 'killed' by going to recents and swiping it away

So what will happen in your case I believe is that when the app starts we'll increment the session count once because a user is registered. Then it will be incremented a second time when you reset and register the user again.

**Comment by slorber** [comment]
@harkin Unfortunatly I never built any android app so your solution is a bit obscure to me :)

If you did only try to increment the counter once `ondeviceready` has been called this would solve this issue problem, but also the problem of the session counter being incremented twice :) But I don't know how to wait for the `ondeviceready` in your Android code :(



```
    @Override protected void pluginInitialize() {
        this.setUpIntercom();
    }

    @Override public void onStart() {
        //We also initialize intercom here just in case it has died. If Intercom is already set up, this won't do anything.
        this.setUpIntercom();

        if (Intercom.client().openGCMMessage(cordova.getActivity().getIntent().getData())) {
            cordova.getActivity().getIntent().setData(null);
        }
    }
```

Maybe the problem is here and only `onStart` should call `setUpIntercom()` ? 

I tried to remove the call in `pluginInitialize()` and it seems the problem does not appear anymore on startup... But actually the widget seems to not be shawn even after being logged in (it seems i have to put the app in background and resume it after 20 seconds otherwise the widget keeps being hidden)

**Comment by slorber** [comment]
I think it may be related to the `onload=true` that is set in `config.xml` of the plugin.

According to what I see here:
https://github.com/apache/cordova-android/blob/4bf705a3d39b34400388265381a9975b246e3779/framework/src/org/apache/cordova/PluginManager.java

What I understand is that onload=true permits to eagerly load the plugin, while otherwise it is somehow "lazyloaded" (I mean only when trying to use the plugin API). I may be wrong.
Maybe the 2nd option is better? don't know

**Comment by jtreanor** [comment]
Hi @slorber,

Just catching up now. Thanks for reporting this. Have you tried setting `onload=false`?

The underlying reason for this issue is that native iOS/Android code has no awareness of the Cordova lifecycle.

**Comment by slorber** [comment]
@jtreanor yes I tried it and it seems I did not get the intercom widget running after ondeviceready but only after going in background and resuming, which is still not good. However I'm not sure my test was correct as I'm not very confortable how to monkeypatch cordova plugins right now :)

**Comment by jtreanor** [comment]
Unfortunately, controlling through the JS isn't an option because the native libraries depend of the application's lifecycle.

Can you explain how a user becomes logged out? Do you automatically log them out on app close?

**Comment by slorber** [comment]
When they log in, they get a token to access the API.
If the "remember me" checkbox is ticked, then this token is persisted, otherwise it is transient.

So yes, when the token is transient and the app is closed/killed/crash, the user is implicitly logged out as on next start we don't have the token anymore

**Comment by jtreanor** [comment]
Yeah, this is going to be a little tricky to work around.

A mandatory "remember me" option would certainly solve this.

You could probably work around it by setting messages to be hidden in `setUpIntercom` but unfortunately there is no nice way to do this.

**Comment by slorber** [comment]
Ok so we finally decided to do like everybody and keep the user connected until explicit logout, and do not let user be disconnected implicitly (app kill/crash) as it does not give opportunity to run any js code

**Comment by jtreanor** [comment]
Closing because this is an unfortunate limitation of the Cordova platform itself.

**Comment by slorber** [comment]
@jtreanor I've just created a Cordova issue for that btw:
https://issues.apache.org/jira/browse/CB-9661

**Comment by jtreanor** [comment]
Thanks @slorber
